### PR TITLE
Fixing ThreadSanitizer data race error in the LIVE VIEW when accessing no_users_thread variable

### PR DESCRIPTION
**Pull request by vzakaznikov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Bug Fix

Fixing ThreadSanitizer data race error in the LIVE VIEW when accessing no_users_thread variable.

### CODE REVIEW

**Review by akuzm — APPROVED** [review]
I have a hunch that a proper fix would involve rethinking the architecture, not adding more mutexes, but have no resources for an in-depth review now. Will merge as is.

**Comment by vzakaznikov** [comment]
> I have a hunch that a proper fix would involve rethinking the architecture, not adding more mutexes, but have no resources for an in-depth review now. Will merge as is.

I totally agree. Adding mutexes right now is a just quick fix. The implementation has to be revisited. Maybe the support for live views that are automatically dropped should be removed as this feature added the usage of an anti-pattern where we are dropping a table in a thread that later can't be joined properly.

File: dbms/src/Storages/LiveView/StorageLiveView.cpp [hunk 1]
```diff
@@ -453,21 +458,28 @@ void StorageLiveView::shutdown()
...
 StorageLiveView::~StorageLiveView()
 {
     shutdown();
-    if (no_users_thread.joinable())
-        no_users_thread.detach();
+
+    {
+        std::lock_guard lock(no_users_thread_mutex);
+        if (no_users_thread.joinable())
+            no_users_thread.detach();
```
Comment (alexey-milovidov) [inline dbms/src/Storages/LiveView/StorageLiveView.cpp:481]: The fix is totally incorrect:

https://clickhouse-test-reports.s3.yandex.net/14404/7ca155c417a4ac64f448c390cc5f5462540b204f/stress_test_(thread)/stderr.log

Calling `std::thread::detach` is just guaranteed error.

Comment (vzakaznikov) [reply dbms/src/Storages/LiveView/StorageLiveView.cpp:481]: I don't think it is totally incorrect but I fully agree that calling and using `std::thread::detach` is bad. So what would be a better way to asynchronously drop table? The use case here is that if `TIMEOUT` is specified during the creation of `LIVE VIEW` we want to drop it automatically if there are no users that are watching this view after some `N` number of seconds.

Comment (alexey-milovidov) [reply dbms/src/Storages/LiveView/StorageLiveView.cpp:481]: Maybe just remove the TIMEOUT feature?

Comment (vzakaznikov) [reply dbms/src/Storages/LiveView/StorageLiveView.cpp:481]: So you think that there is nothing in ClickHouse where we could schedule a callback?

**Comment by vzakaznikov** [comment]
Maybe we should do something like this https://github.com/ClickHouse/ClickHouse/blob/83a89fe858b03290b3c70eb0a6eb0e9145e40018/src/Storages/RabbitMQ/StorageRabbitMQ.cpp#L105?

File: dbms/src/Storages/LiveView/StorageLiveView.cpp (hunk 1 above)
Comment (vzakaznikov) [reply dbms/src/Storages/LiveView/StorageLiveView.cpp:481]: Maybe we should do something like this? https://github.com/ClickHouse/ClickHouse/blob/83a89fe858b03290b3c70eb0a6eb0e9145e40018/src/Storages/RabbitMQ/StorageRabbitMQ.cpp#L105

Comment (alexey-milovidov) [reply dbms/src/Storages/LiveView/StorageLiveView.cpp:481]: Yes, it can be Ok.

Comment (vzakaznikov) [reply dbms/src/Storages/LiveView/StorageLiveView.cpp:481]: Ok, perfect. I will try it out and create a PR.
